wandb/apis/importers/wandb.py

```diff
@@ -0,0 +1,1133 @@
+
+    def cpus_used(self) -> Optional[int]:  # can we get the model?
+        ...
+
+    def memory_used(self) -> Optional[int]:
+        ...
+
+    def runtime(self) -> Optional[int]:
+        wandb_runtime = self.run.summary.get("_wandb", {}).get("runtime")
+        base_runtime = self.run.summary.get("_runtime")
+
+        t = coalesce(wandb_runtime, base_runtime)
+        if t is None:
+            return t
+        return int(t)
+
+    def start_time(self) -> Optional[int]:
+        t = dt.fromisoformat(self.run.created_at).timestamp()
+        return int(t)
+
+    def code_path(self) -> Optional[str]:
+        fname = self._find_in_files("wandb-metadata.json")
+        if fname is None:
+            return None
+
+        with open(fname) as f:
+            result = json.loads(f.read())
+            return "code/" + result.get("codePath", "")
+
+    def cli_version(self) -> Optional[str]:
+        fname = self._find_in_files("config.yaml")
+        if fname is None:
+            return None
+
+        with open(fname) as f:
+            result = yaml.safe_load(f)
+            return result.get("_wandb", {}).get("value", {}).get("cli_version")
+
+    def files(self) -> Optional[Iterable[Tuple[str, str]]]:
+        files_dir = f"./wandb-importer/{self.run_id()}/files"
+        if self._files is not None:
+            yield from self._files
+            return
+
+        self._files = []
+        for f in self.run.files():
+            try:
+                result = f.download(files_dir, exist_ok=True)
+            except Exception as e:
+                wandb_logger.error(
+                    f"Error downloading file ({f}) -- {e}",
+                    extra={
+                        "entity": self.entity(),
+                        "project": self.project(),
+                        "run_id": self.run_id(),
+                    },
+                )
+                continue
+            else:
+                file_and_policy = (result.name, "now")
+                self._files.append(file_and_policy)
+                yield file_and_policy
+
+    def logs(self) -> Optional[Iterable[str]]:
+        fname = self._find_in_files("output.log")
+        if fname is None:
+            return
+
+        with open(fname) as f:
+            yield from f.readlines()
+
+    def _modify_table_artifact_paths(self, row: Dict[str, Any]) -> Dict[str, Any]:
+        table_keys = []
+        for k, v in row.items():
+            if (
+                isinstance(v, (dict, wandb.old.summary.SummarySubDict))
+                and v.get("_type") == "table-file"
+            ):
+                table_keys.append(k)
+
+        for table_key in table_keys:
+            obj = row[table_key]["artifact_path"]
+            obj_name = obj.split("/")[-1]
+            art_path = f"{self.entity()}/{self.project()}/run-{self.run_id()}-{table_key}:latest"
+            art = None
+
+            # Try to pick up the artifact within 20 seconds
+            for _ in range(10):
```

Comment:
Why do we need this retry?

Reply:
Sometimes this artifact does not appear immediately.  I'm not exactly sure why, but I think it's related to being a table artifact